### ISSUE TITLE
Prior predictive sampling with pm.Bound gives wrong results

### ISSUE THREAD

**Issue by ricardoV94** [pymc-devs/pymc#4643]
Using `pm.Bound` for distributions whose arguments include other model parameters (instead of constants) will lead to a model that is wrongly sampled in `sample_prior_predictive()`

```python
with pm.Model() as m:
    pop = pm.Normal('pop', 2, 1)
    ind = pm.Bound(pm.Normal, lower=-2, upper=2)('ind', mu=pop, sigma=.5)
    
    prior = pm.sample_prior_predictive()
    trace = pm.sample()
```

The samples for the bounded distribution are fine (i.e., one obtains the same results if doing prior_predictive or sampling without data):

```python
sns.kdeplot(prior['ind'])
sns.kdeplot(trace['ind'])
```
![image](https://user-images.githubusercontent.com/28983449/114866404-cbfe8180-9df3-11eb-9c12-c4fd5d88e9ae.png)

But the samples for the hyper-parameters do not match:
```python
sns.kdeplot(prior['pop'])
sns.kdeplot(trace['pop'])
``` 
![image](https://user-images.githubusercontent.com/28983449/114866371-c143ec80-9df3-11eb-88fe-280fe7f2b729.png)

What is going on? I think the way `pm.Bound` is implemented, it corresponds to adding an arbitrary factor to the model logp of the kind:

```python
pm.Potential('err', pm.math.switch((ind >= -2) & (ind <= 2), 0, -np.inf))
```
Which obviously is not (and cannot) be accounted for in predictive sampling.

**Comment by OriolAbril** [comment]
This could be helpful in implementing a fix: https://mc-stan.org/docs/2_18/stan-users-guide/truncated-random-number-generation.html

**Comment by ricardoV94** [comment]
The thing is that Bound does not really correspond to a Truncated distribution. If it did, it would not affect the prior of the hyperparameters during (correctly implemented) prior predictive sampling.

**Comment by ricardoV94** [comment]
The `TruncatedNormal` which is a properly truncated distribution does not affect the prior predictive of its hyperparameters

```python
with pm.Model() as m:
    pop = pm.Normal('pop', 2, 1)
    ind = pm.TruncatedNormal('ind', lower=-2, upper=2, mu=pop, sigma=.5)
    prior = pm.sample_prior_predictive()
    trace = pm.sample()
```
The prior for the Truncated is different, note the smaller left tail
```python
sns.kdeplot(prior['ind'])
sns.kdeplot(trace['ind'])
```
![image](https://user-images.githubusercontent.com/28983449/114869219-07e71600-9df7-11eb-84c1-0abde0812cf3.png)

The prior for the hyperparameter is not affected by the Truncation downstream
```python
sns.kdeplot(prior['pop'])
sns.kdeplot(trace['pop'])
```
![image](https://user-images.githubusercontent.com/28983449/114869195-01f13500-9df7-11eb-88db-af8204606a84.png)

**Comment by ricardoV94** [comment]
@OriolAbril your point made me recheck the Bound random method and I am very unsure as to whether it's logic is sound (although I couldn't confirm it's not either):

https://github.com/pymc-devs/pymc3/blob/2dee9dafb3541043ae23b66fe9262e70fbd2769a/pymc3/distributions/bound.py#L90-L95

In particular it ignores the order of the `point` values. It seems that if a `point` happens to include a parameter value that is easy enough to sample from within the bound constraints and 99 values that do not, it will happy resample 99 times from the point of high bounded likelihood. I am not completely sure this is violates the statistical behavior of Bound though.

**Comment by michaelosthege** [comment]
@ricardoV94 was this `v3` or `v4`? Please set labels accordingly.

**Comment by ricardoV94** [comment]
It's an issue in V3. Bound hasn't been refactored to V4 but will be the same. Anyway no point in changing things for V3 so I'll put a label accordingly
